src/webgpu/shader/execution/expression/expression.ts

```diff
@@ -496,14 +507,86 @@ export function basicExpressionBuilder(expressionBuilder: ExpressionBuilder): Sh
+        //   const sign_bit : u32 = select(0, 0x80000000, ${expressionBuilder(map(c.input, v => v.wgsl()))} < 0);
+        //
+        //   // The exponent is stored in bits [20..31] of high_u32, special case for zero
+        //   const exponent_bits : u32 = select(u32(f.exp + exponent_bias) << 20, 0, is_zero);
+        //
+        //   // Extract the high 20 bits of the mantissa as u32, special case for zero
+        //   const high_mantissa_bits : u32 = select(u32(f.fract * (1 << 21)) & ((1 << 20) - 1), 0, is_zero);
+        //
+        //   // Subtract out the high 20 bits from the mantissa.
+        //   // const low_mantissa = f.fract - floor(f.fract * (1 << 20)) / (1 << 20);
+        //
+        //   // Extract the low 32 bits of the mantissa as u32
+        //   // The lower mantissa may or may not have a leading one bit that needs to be masked.
+        //   // Unfortunately there is no simple way to bitmask an AbstractInt, because it needs to be first to be
+        //   // concretized, and if that bit is present the u32 cast will fail due to being out of range.
+        //   // So to remove it (assuming there are no higher bits), without bitcasting, one can subtract (1 << 32).
+        //   // But this does not work if the bit isn't present, because that returns are negative value, which cannot
+        //   // then be cast to u32.
+        //   // So the snippet calculates both values as AbstractInts, and then selects between them.
+        //   const adjusted_low_mantissa_bits = low_mantissa * (1 << 53) - 0x100000000;
+        //   const unadjusted_low_mantissa_bits = low_mantissa * (1 << 53);
+        //   const low_mantissa_bits = u32(select(adjusted_low_mantissa_bits, unadjusted_low_mantissa_bits, adjusted_low_mantissa_bits  < 0));
+        //
+        //   // Low bits are just the 32 low bits of the mantissa, or just zero for +/-0
+        //   outputs[${i}].value.x = select(low_mantissa_bits, 0, is_zero);
+        //   // high bits are the sign, the exponent and the remaining mantissa bits
+        //   outputs[${i}].value.y = sign_bit | exponent_bits | high_mantissa_bits;
+        body = cases
+          .map(
+            // prettier-ignore
+            (c, i) => `  {
+    const exponent_bias = 1022;
+    const f = frexp(abs(${expressionBuilder(map(c.input, v => v.wgsl()))}));
+    const is_zero : bool = (f.exp == 0) && (f.fract == 0.0);
+    const_assert(is_zero || (f.exp != 0)); // Subnormals not currently supported, but zero is
+    const sign_bit : u32 = select(0, 0x80000000, ${expressionBuilder(map(c.input, v => v.wgsl()))} < 0);
+    const exponent_bits : u32 = select(u32(f.exp + exponent_bias) << 20, 0, is_zero);
+    const high_mantissa_bits : u32 = select(u32(f.fract * (1 << 21)) & ((1 << 20) - 1), 0, is_zero);
+    const low_mantissa = f.fract - floor(f.fract * (1 << 20)) / (1 << 20);
+    const adjusted_low_mantissa_bits = low_mantissa * (1 << 53) - 0x100000000;
```

Comment:
The fact that I have to select between adjust/not adjust here is the part that I wasn't expecting/would like someone to take a look at.

I will provide some examples of why this is needed in  a comment on this PR